### PR TITLE
Extract FixPath logic from Value

### PR DESCRIPTION
**Pull request by fsteeg**
As discussed in our meeting, this is an interim state for #102.

### CODE REVIEW

File: metafix/src/main/java/org/metafacture/metafix/FixMethod.java [hunk 1]
```diff
@@ -355,7 +355,7 @@ public void apply(final Metafix metafix, final Record record, final List<String>
     downcase {
         @Override
         public void apply(final Metafix metafix, final Record record, final List<String> params, final Map<String, String> options) {
-            record.transformFields(params, String::toLowerCase);
+            new FixPath(params.get(0)).transformIn(record, s -> s.toLowerCase());
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/FixMethod.java:358]: Makes me a little sad :cry: 

FTR:

```
error: reference to transformIn is ambiguous

both method transformIn(org.metafacture.metafix.Value.Hash,java.util.function.BiConsumer<org.metafacture.metafix.Value.TypeMatcher,java.util.function.Consumer<org.metafacture.metafix.Value>>) in org.metafacture.metafix.FixPath and method transformIn(org.metafacture.metafix.Value.Hash,java.util.function.UnaryOperator<java.lang.String>) in org.metafacture.metafix.FixPath match
```

File: metafix/src/main/java/org/metafacture/metafix/FixPath.java [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.metafacture.metafix;
+
+import org.metafacture.metafix.Value.Array;
+import org.metafacture.metafix.Value.Hash;
+import org.metafacture.metafix.Value.ReservedField;
+import org.metafacture.metafix.Value.TypeMatcher;
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/FixPath.java:22]: Hm, I'm undecided if I like these `import`s of nested types; in particular as they refer to a sibling class. Anyway, it's easier for now not to change all references.

File: metafix/src/main/java/org/metafacture/metafix/FixPath.java [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+        final String field = path[0];
+        if (field.equals(ASTERISK)) {
+            // TODO: search in all elements of value.asHash()?
+            return new FixPath(tail(path)).findIn(hash);
+        }
+        return path.length == 1 || !hash.containsField(field) ? hash.get(field) :
+            findNested(hash, field, tail(path));
+    }
+
+    /*package-private*/ Value findIn(final Value.Array array) {
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/FixPath.java:59]: Except for this one. Should be changed to `Array` then.

File: metafix/src/main/java/org/metafacture/metafix/FixPath.java [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+        else {
+            for (int i = 0; i < size; ++i) {
+                transformValueAt(array, i, tail(path), operator);
+            }
+        }
+
+        array.removeIf(v -> Value.isNull(v));
+    }
+
+    protected enum InsertMode {
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/FixPath.java:178]: Should be `private`?

File: metafix/src/main/java/org/metafacture/metafix/Value.java [hunk 5]
```diff
@@ -256,11 +255,8 @@ public String toString() {
         return Arrays.copyOfRange(fields, 1, fields.length);
     }
 
-    private void transformFields(final String[] fields, final UnaryOperator<String> operator) {
-        matchType()
-            .ifArray(a -> a.transformFields(fields, operator))
-            .ifHash(h -> h.transformFields(fields, operator))
-            .orElseThrow();
+    static String[] split(final String fieldPath) {
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/Value.java:258]: `/*package-private*/`

File: metafix/src/main/java/org/metafacture/metafix/Value.java [hunk 6]
```diff
@@ -274,7 +270,7 @@ private void transformFields(final String[] fields, final UnaryOperator<String>
         private final Set<Type> expected = new HashSet<>();
         private final Value value;
 
-        private TypeMatcher(final Value value) {
+        /*package-private*/ TypeMatcher(final Value value) {
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/Value.java:273]: Should be `private`.

File: metafix/src/main/java/org/metafacture/metafix/Value.java [hunk 7]
```diff
@@ -705,45 +533,14 @@ public Value append(final String fieldPath, final String newValue) {
          */
         public Value get(final String field) {
             // TODO: special treatment (only) for exact matches?
-            final List<Value> list = findFields(field).map(map::get).collect(Collectors.toList());
+            final List<Value> list = matchFields(field, Stream::filter).map(map::get).collect(Collectors.toList());
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/Value.java:536]: Why did you drop the `findFields()` helper?

File: metafix/src/main/java/org/metafacture/metafix/FixPath.java [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+                    hash.remove(f);
+
+                    if (operator != null) {
+                        value.matchType()
+                            .ifString(s -> new FixPath(f).appendIn(hash, operator.apply(s)))
+                            .orElseThrow();
+                    }
+                }
+                else {
+                    new TypeMatcher(value)
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/FixPath.java:144]:
```suggestion
                    value.matchType()
```

File: metafix/src/main/java/org/metafacture/metafix/FixPath.java [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+    private String[] tail(final String[] fields) {
+        return Arrays.copyOfRange(fields, 1, fields.length);
+    }
+
+    private void transformValueAt(final Array array, final int index, final String[] p, final UnaryOperator<String> operator) {
+        final Value value = array.get(index);
+        if (value != null) {
+            value.matchType()
+                .ifString(s -> array.set(index, operator != null ? new Value(operator.apply(s)) : null))
+                .orElse(v -> new Value.TypeMatcher(v)
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/FixPath.java:279]:
```suggestion
                .orElse(v -> v.matchType()
```

File: metafix/src/main/java/org/metafacture/metafix/FixPath.java [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+        if (oldValue != null) {
+            final Value newValue = oldValue.extractType(consumer);
+
+            if (newValue != null) {
+                new FixPath(path).insertInto(hash, InsertMode.REPLACE, newValue);
+            }
+        }
+    }
+
+    /*package-private*/ void transformIn(final Hash hash, final UnaryOperator<String> operator) {
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/FixPath.java:120]: Must be `public`. Currently needed in Limetrans. Or what is the intended replacement for `Record.transformFields(List<String>, UnaryOperator<String>)`?

**Review by blackwinter — CHANGES_REQUESTED** [review]
Overall, nice to see `Value` shrink a little in size :)

File: metafix/src/main/java/org/metafacture/metafix/Value.java (hunk 7 above)
Comment (fsteeg) [reply metafix/src/main/java/org/metafacture/metafix/Value.java:536]: I mostly tried to trim down the similar named methods, so I dropped one-liners. Feel free to add anything back that you feel is an improvement.

File: metafix/src/main/java/org/metafacture/metafix/Value.java (hunk 6 above)
Comment (blackwinter) [reply metafix/src/main/java/org/metafacture/metafix/Value.java:273]: Not if you switch to `Value.matchType()`. See suggestions below.

Comment (fsteeg) [reply metafix/src/main/java/org/metafacture/metafix/Value.java]: Right, saw it a moment after I commented (and deleted my comment).

**Comment by fsteeg** [comment]
Thanks for the review! Addressed your feedback in https://github.com/metafacture/metafacture-fix/pull/133/commits/a9aac45236e56c10d67d3b584fd52df15f95c814.

**Review by blackwinter — APPROVED** [review]
Great, thanks!

**Pull request merged**